src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -7,82 +7,60 @@ import (
 )
 
 ////////////////////////////////////////////////////////////////////////////////
 // Actor methods
 ////////////////////////////////////////////////////////////////////////////////
 
-func (a *StorageMarketActorCode_I) WithdrawBalance(rt Runtime, minerAddr addr.Address, amountRequested actor.TokenAmount) {
+func (a *StorageMarketActorCode_I) WithdrawBalance(rt Runtime, entryAddr addr.Address, amountRequested actor.TokenAmount) {
 	IMPL_FINISH() // BigInt arithmetic
 	amountSlashedTotal := actor.TokenAmount(0)
 
-	TODO() // TODO: must also handle case of non-miner (client) sender
-
-	ownerAddr, workerAddr := _rtGetMinerAccountsAssert(rt, minerAddr)
-	rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
-	TODO() // TODO: should workerAddr be permitted here? Depends on actor principal security assumptions.
-
-	callerAddr := rt.ImmediateCaller()
-
 	if amountRequested < 0 {
-		rt.AbortArgMsg("sma.WithdrawBalance: negative amount.")
+		rt.AbortArgMsg("Negative amount.")
 	}
 
-	h, st := a.State(rt)
+	recipientAddr := _rtValidateImmediateCallerDetermineRecipient(rt, entryAddr)
 
-	if !st._addressEntryExists(minerAddr) {
-		rt.AbortArgMsg("sma.WithdrawBalance: address entry does not exist")
-	}
+	h, st := a.State(rt)
+	st._rtAbortIfAddressEntryDoesNotExist(rt, entryAddr)
 
 	// Before any operations that check the balance tables for funds, execute all deferred
 	// deal state updates.
 	//
 	// Note: as an optimization, implementations may cache efficient data structures indicating
 	// which of the following set of updates are redundant and can be skipped.
-	amountSlashedTotal += st._rtUpdatePendingDealStatesForParty(rt, minerAddr)
+	amountSlashedTotal += st._rtUpdatePendingDealStatesForParty(rt, entryAddr)
 
-	minBalance := st._getLockedReqBalanceInternal(minerAddr)
+	minBalance := st._getLockedReqBalanceInternal(entryAddr)
 	newTable, amountExtracted, ok := actor.BalanceTable_WithExtractPartial(
-		st.EscrowTable(), minerAddr, amountRequested, minBalance)
+		st.EscrowTable(), entryAddr, amountRequested, minBalance)
 	Assert(ok)
 	st.Impl().EscrowTable_ = newTable
 
 	UpdateRelease(rt, h, st)
 
-	// send funds to caller
-	rt.SendPropagatingErrors(&vmr.InvocInput_I{
-		To_:    callerAddr,
-		Value_: amountExtracted,
-	})
-
 	rt.SendFunds(addr.BurntFundsActorAddr, amountSlashedTotal)
+	rt.SendFunds(recipientAddr, amountExtracted)
 }
 
-func (a *StorageMarketActorCode_I) AddBalance(rt Runtime, minerAddr addr.Address) {
-	TODO() // TODO: must also handle case of non-miner (client) sender
-
-	ownerAddr, workerAddr := _rtGetMinerAccountsAssert(rt, minerAddr)
-	rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
-	TODO() // TODO: should workerAddr be permitted here? Depends on actor principal security assumptions.
-
-	msgValue := rt.ValueReceived()
+func (a *StorageMarketActorCode_I) AddBalance(rt Runtime, entryAddr addr.Address) {
+	_rtValidateImmediateCallerDetermineRecipient(rt, entryAddr)
```

Comment:
In conversation we clearly decided that anyone could add balance to an account. This should not do any caller verification at all.

Reply:
(Addressed via out-of-band discussions.)